diff --git a/frontend/index.js b/frontend/index.js
index 42b9716..5449938 100644
--- a/frontend/index.js
+++ b/frontend/index.js
@@ -11,7 +11,7 @@ btn.addEventListener("click", () => {
     xhr.onreadystatechange = function () {
         if (xhr.readyState === 4 && xhr.status === 200) {
             var json = JSON.parse(xhr.responseText);
-            document.getElementById("label").innerHTML = json.shortened_url;
+            document.getElementById("label").innerHTML = '<a href=\'' + json.shortened_url + '\'>' + json.shortened_url + '</a>';
         }
     };
     xhr.send(dataJson);
diff --git a/server.js b/server.js
index 930cd81..e04949d 100644
--- a/server.js
+++ b/server.js
@@ -1,16 +1,14 @@
 const express = require('express');
 const path = require('path');
+const JSONdb = require('simple-json-db');
 const app = express();
+const db = new JSONdb('./map.json');
 const port = 9000;
 
 app.use(express.json());
 app.use(express.static(path.join(__dirname, 'frontend')));
 
-function getRand(lim) {
-	return Math.floor(Math.random() * lim);
-}
-
-function genShortenedUrl(url) {
+function genShortenedUrl() {
 	let url_ = '';
 	let possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
 	for(let x = 1; x <= 6; x++) {
@@ -19,19 +17,41 @@ function genShortenedUrl(url) {
 	return 'http://127.0.0.1:' + port.toString() + '/' + url_;
 }
 
+function map_(url) {
+	let shortened_url = '';
+	let flag = false;
+	for (i in db.JSON()) {
+		if(db.get(i) == url) {
+			shortened_url = i;
+			flag = true;
+		}
+	}
+	
+	if(!flag) {
+		shortened_url = genShortenedUrl();
+	}
+
+	db.set(shortened_url, url);
+	return shortened_url;
+}
+
 app.get('/:id', function(req, res){
-	res.send('The id you specified is ' + req.params.id);
+	shortened_url_ = 'http://127.0.0.1:' + port + '/' + req.params.id;
+	let exists_ = db.has(shortened_url_);
+	if(exists_) {
+		res.redirect(db.get(shortened_url_));
+	}
+	else {
+		res.send('<h2>Error 404! URL doesn\'t exist</h2>');
+	}
 });
 
 app.get('/', (req, res) => {
 	res.sendFile('index.html')
-	// res.send('<h2>Hello from Node JS</h2>');
 });
 
 app.post('/', (req, res) => {
-	console.log(req.body);
-	shortened_url = genShortenedUrl('');
-	console.log(shortened_url);
+	shortened_url = map_(req.body.url);
 	res.send("{\"shortened_url\" : \"" + shortened_url + "\"}");
 })
 
